README.md
# dominion-builder

## Getting Started

```
$ yarn install && yarn pip
$ yarn dev
```

Access http://localhost:3000


CHANGELOG.md
# Change Log

## 2022.04.16 Version 0.2.1

### 🔧 Fix
* Support split piles and travelers
* Add a concept image and support Twitter OGP

## 2022.04.15 Version 0.2.0

### 🚀 New
* Add Change Log.
* Add the function to specify room parameters (# of users and cards).
* Server now returns voting results.

## 2022.04.13 Version 0.1.1

### 🔧 Fix
* Update the list of cards.

## 2022.04.12 Version 0.1.0

### 🚀 New
* Project launched!
* Add the function to create rooms.
* Add the function to select cards.

--- template ---

### 🚀 New
*

### 🔧 Fix
*

### 🧹 Cleanup
*

pages/index.js
import { useState, useEffect } from 'react'
import {
  Box,
  Button,
  Heading,
  HStack,
  Image,
  Link,
  Select,
  Spacer,
  Text,
  useToast,
} from '@chakra-ui/react'
import { useRouter } from 'next/router'
import { v4 as uuidv4 } from 'uuid'

import { Layout } from '../components/layout'

const backendURL = (
  process.env.BACKEND_URL ? process.env.BACKEND_URL : 'ws://localhost:8000'
).replace('ws', 'http')

export default function Home() {
  const [isCreatingRoom, setIsCreatingRoom] = useState(false)
  const [nUsers, setNUsers] = useState(null)
  const [nSelections, setNSelections] = useState(null)

  const router = useRouter()
  const toast = useToast()

  const createRoom = () => {
    if (!nUsers || !nSelections) {
      toast({
        title: '新しく部屋を作るには',
        description: '参加人数とプレイヤーが指定する枚数を決めてください。',
        status: 'info',
        duration: 5000,
        isClosable: true,
      })
      return
    }

    if (nUsers * nSelections > 10) {
      toast({
        title: '指定するカードが多すぎます',
        description: '人数 ✕ 枚数を10以下にしてください。',
        status: 'error',
        duration: 5000,
        isClosable: true,
      })
      return
    }

    let trial = 0
    setIsCreatingRoom(true)

    const timer = setInterval(() => {
      fetch(backendURL)
        .then((res) => {
          clearInterval(timer)
          const encodedRoomParameter = (nUsers - 2) * 3 + (nSelections - 1)
          router.push(
            `/${'' + encodedRoomParameter + uuidv4().substring(0, 3)}/`
          )
        })
        .catch((error) => {
          trial += 1
          if (trial > 9) {
            clearInterval(timer)
            setIsCreatingRoom(false)
            toast({
              title: '部屋の作成に失敗しました',
              description: 'もう一度試してみてください。',
              status: 'error',
              duration: 5000,
              isClosable: true,
            })
          }
        })
    }, 1000)
  }

  useEffect(() => {
    fetch(backendURL)
      .then((res) => {})
      .catch((error) => {})
  }, [])

  return (
    <Layout>
      <Heading size="xl" mb={2}>
        Dominion Builder
      </Heading>
      <Text fontSize="md" mb={2}>
        ドミニオンのサプライをプレイヤーの投票によって決めるWebサイトです。部屋ごとにリンクが発行されるので、他の人を簡単に招待することができます。
      </Text>
      <Image mb={2} src="/concept.png" />
      <Box p={4} mb={6} bgColor="gray.50" borderRadius={6}>
        <Text fontSize="md" mb={3}>
          参加人数とプレイヤーが指定するカード枚数を決めて部屋を作りましょう。
        </Text>
        <HStack>
          <Select
            placeholder="人数"
            flex="1"
            onChange={(e) => {
              setNUsers(e.target.value)
            }}
          >
            {[2, 3, 4].map((n) => (
              <option key={`n-players-${n}`} value={n}>
                {n}人
              </option>
            ))}
          </Select>
          <Select
            placeholder="枚数"
            flex="1"
            onChange={(e) => {
              setNSelections(e.target.value)
            }}
          >
            {[1, 2, 3].map((n) => (
              <option key={`n-selections-${n}`} value={n}>
                {n}枚
              </option>
            ))}
          </Select>
          <Button
            onClick={createRoom}
            isLoading={isCreatingRoom}
            loadingText="部屋を作成中"
            colorScheme="blue"
            width="160px"
          >
            新しく部屋を作る
          </Button>
        </HStack>
      </Box>
      <HStack spacing={4}>
        <Link
          onClick={() => {
            toast({
              title: '他の人が作った部屋に入るには',
              description:
                '部屋のリンクをブラウザのアドレスバーに入力してください。',
              status: 'info',
              duration: 5000,
              isClosable: true,
            })
          }}
        >
          部屋に入る
        </Link>
        <Link
          href={
            'https://github.com/shoya140/dominion-builder/blob/main/CHANGELOG.md'
          }
        >
          Change Log
        </Link>
        <Link href={'https://github.com/shoya140/dominion-builder/'}>
          GitHub
        </Link>
        <Spacer />
      </HStack>
    </Layout>
  )
}


backend/main.py
import json
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.middleware.cors import CORSMiddleware


class ConnectionManager:
    def __init__(self):
        self.connections = {}

    async def connect(self, websocket: WebSocket, room_id: str, user_id: str):
        await websocket.accept()
        if not room_id in self.connections:
            self.connections[room_id] = {}
        self.connections[room_id][user_id] = {
            "user_name": user_id,
            "websocket": websocket,
            "selections": []
        }
        await self.broadcast_room_status(room_id)

    async def disconnect(self, room_id: str, user_id: str):
        del self.connections[room_id][user_id]
        if len(self.connections[room_id]) == 0:
            del self.connections[room_id]
            return  # do not broadcast if the room is empty
        await self.broadcast_room_status(room_id)

    async def handle_message(self, room_id: str, user_id, message: str):
        data = json.loads(message)

        if data["event_type"] == "submit selections":
            self.connections[room_id][user_id]["selections"] = data["selections"]
            n_users = int(room_id[0]) // 3 + 2  # defined in the frontend
            if len([x for x in self.connections[room_id].values() if len(x["selections"]) > 0]) == (n_users):
                await self.broadcast_results(room_id)

        if data["event_type"] == "update user name":
            self.connections[room_id][user_id]["user_name"] = data["user_name"]

        await self.broadcast_room_status(room_id)

    async def broadcast_room_status(self, room_id):
        filtered_user_status = [
            {
                "userId": user_id,
                "userName": x["user_name"],
                "submitted": len(x["selections"])
            } for user_id, x in self.connections[room_id].items()
        ]
        payload = {
            "eventType": "users updated",
            "users": filtered_user_status
        }
        await self.broadcast(json.dumps(payload), room_id)

    async def broadcast_results(self, room_id: str):
        cards = {}
        logs = []
        for user in self.connections[room_id].values():
            selections = "と".join(["「" + x + "」" for x in user["selections"]])
            logs.append(user["user_name"] + "は" + selections + "を指定しました。")
            for selection in user["selections"]:
                cards[selection] = 1
        payload = {
            "eventType": "voting completed",
            "cards": list(cards.keys()),
            "logs": logs
        }
        await self.broadcast(json.dumps(payload), room_id)

    async def broadcast(self, message: str, room_id: str):
        try:
            for user in self.connections[room_id].values():
                await user["websocket"].send_text(message)
        except:
            print("warning: The connection was already closed")


manager = ConnectionManager()

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
def read_root():
    clients = {}
    for room, users in manager.connections.items():
        clients[room] = list(users.keys())
    return clients


@app.websocket("/{room_id}/{user_id}")
async def websocket_endpoint(websocket: WebSocket, room_id: str, user_id: str):
    await manager.connect(websocket, room_id, user_id)

    try:
        while True:
            message = await websocket.receive_text()
            await manager.handle_message(room_id, user_id, message)
    except WebSocketDisconnect:
        await manager.disconnect(room_id, user_id)


components/layout.js
import Head from 'next/head'
import { useRouter } from 'next/router'

const frontendURL = process.env.FRONTEND_URL
  ? process.env.FRONTEND_URL
  : 'http://localhost:3000'

export function Layout({ children, wide }) {
  const router = useRouter()

  return (
    <div>
      <Head>
        <title>Dominion Builder</title>
        <meta
          name="description"
          content="ドミニオンのサプライをプレイヤーの投票によって決めるWebサイトです。部屋ごとにリンクが発行されるので、他の人を簡単に招待することができます。"
        />
        <meta name="twitter:card" content="summary" />
        <meta property="og:url" content={`${frontendURL}${router.asPath}`} />
        <meta property="og:title" content="Dominion Builder" />
        <meta
          property="og:description"
          content="ドミニオンのサプライをプレイヤーの投票によって決めるWebサイトです。"
        />
        <meta property="og:image" content={`${frontendURL}/ogp.png`} />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <div className="container">
        <div className={wide ? 'content-wide' : 'content'}>{children}</div>
      </div>
    </div>
  )
}

export function FloatingTop({ children }) {
  return <div className="floating-top">{children}</div>
}


pages/[roomId].js
import { useState, useEffect } from 'react'
import { useRouter } from 'next/router'
import { v4 as uuidv4 } from 'uuid'
import Mikan from 'mikanjs'
import {
  Avatar,
  AvatarBadge,
  Box,
  Button,
  HStack,
  Input,
  Link,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalBody,
  Text,
  Tooltip,
  VStack,
  useClipboard,
  useToast,
  SimpleGrid,
  Center,
  Spacer,
  UnorderedList,
  ListItem,
} from '@chakra-ui/react'
import { LinkIcon } from '@chakra-ui/icons'

import { Layout, FloatingTop } from '../components/layout'
import cards from '../lib/cards.json'

const frontendURL = process.env.FRONTEND_URL
  ? process.env.FRONTEND_URL
  : 'http://localhost:3000'

const backendURL = process.env.BACKEND_URL
  ? process.env.BACKEND_URL
  : 'ws://localhost:8000'

const defaultUserName = process.env.NODE_ENV === 'development' ? 'dev' : ''

const groupedCards = cards.reduce((result, card) => {
  result[card.expansion] = [...(result[card.expansion] || []), card]
  return result
}, {})

let ws

export default function Room() {
  const [userName, setUserName] = useState(defaultUserName)
  const [userNameValue, setUserNameValue] = useState('')
  const [selections, setSelections] = useState([])
  const [submitted, setSubmitted] = useState(false)
  const [users, setUsers] = useState([])
  const [inviteLink, setInviteLink] = useState('')
  const [nUsers, setNUsers] = useState(0)
  const [nSelections, setNSelections] = useState(0)
  const [result, setResult] = useState(null)

  const router = useRouter()
  const toast = useToast()
  const { hasCopied, onCopy } = useClipboard(inviteLink)

  const submitSelections = () => {
    setSubmitted(true)
    ws.send(
      JSON.stringify({
        event_type: 'submit selections',
        selections: selections,
      })
    )
    toast({
      title: '投票を受け付けました',
      description: '他の人の投票完了をお待ち下さい。',
      status: 'success',
      duration: 6000,
      isClosable: true,
    })
  }

  const selectCard = (series) => {
    if (
      submitted ||
      (selections.length === nSelections && !selections.includes(series))
    ) {
      return
    }
    if (selections.includes(series)) {
      setSelections([...selections].filter((s) => s !== series))
    } else {
      setSelections([...selections, series])
    }
  }

  const handleUserNameValueChange = (e) => {
    setUserNameValue(e.target.value)
  }

  const updateUserName = () => {
    setUserName(userNameValue)
    ws.send(
      JSON.stringify({
        event_type: 'update user name',
        user_name: userNameValue,
      })
    )
  }

  const clearUserName = () => {
    setUserName('')
    setUserNameValue('')
  }

  useEffect(() => {
    if (!router.isReady) {
      return
    }

    const encodedRoomParameter = parseInt(router.query.roomId.charAt(0))
    setNUsers(Math.floor(encodedRoomParameter / 3) + 2)
    setNSelections((encodedRoomParameter % 3) + 1)

    const userId = uuidv4().substring(0, 8)
    setInviteLink(`${frontendURL}/${router.query.roomId}`)
    ws = new WebSocket(`${backendURL}/${router.query.roomId}/${userId}`)
    ws.onmessage = (event) => {
      const data = JSON.parse(event.data)
      if (data.eventType === 'users updated') {
        setUsers(data.users)
      }
      if (data.eventType === 'voting completed') {
        console.log(data)
        const c = cards
          .filter(({ name }) => data.cards.includes(name))
          .reduce((result, card) => {
            result[card.expansion] = [...(result[card.expansion] || []), card]
            return result
          }, {})
        setResult({ cards: c, logs: data.logs })
      }
    }
    ws.onopen = () => {}
  }, [router.isReady])

  return (
    <>
      {userName && (
        <>
          {!submitted && selections.length === nSelections && (
            <FloatingTop>
              <Button colorScheme="blue" onClick={submitSelections} size="lg">
                投票する
              </Button>
            </FloatingTop>
          )}
          <Layout wide>
            <Text fontSize="xl" fontWeight="600" textAlign="center">
              Room ID: {router.query.roomId} ({nUsers}人, {nSelections}枚)
            </Text>
            <HStack mt={10} mb={6} overflowX="scroll">
              {users.map(({ userName, submitted }, index) => (
                <Tooltip
                  key={`user-${index}`}
                  hasArrow
                  label={submitted ? '投票完了' : '投票中'}
                  placement="top"
                  bg={submitted ? 'green.400' : 'blue.400'}
                  shadow={false}
                >
                  <Box p={2} borderRadius={4} textAlign="center" width="90px">
                    <Avatar>
                      <AvatarBadge
                        boxSize="1.25em"
                        bg={submitted ? 'green.400' : 'blue.400'}
                      />
                    </Avatar>
                    <Text fontSize="sm" mt={2} isTruncated>
                      {userName}
                    </Text>
                  </Box>
                </Tooltip>
              ))}
              {users.length < nUsers && (
                <Box
                  p={5}
                  borderRadius={4}
                  textAlign="center"
                  onClick={onCopy}
                  bgColor="gray.50"
                  cursor="pointer"
                >
                  <LinkIcon w={6} h={6} />
                  <Text fontSize="xs" mt={2}>
                    {hasCopied ? 'コピーしました' : 'リンクをコピー'}
                  </Text>
                </Box>
              )}
            </HStack>
            <VStack spacing={2} mb={6} align="left">
              {Object.keys(groupedCards).map((expansion) => (
                <div key={`expansion-${expansion}`}>
                  <Text
                    mb={2}
                    fontSize="xl"
                    fontWeight="600"
                    opacity={selections.length === nSelections ? 0.4 : 1}
                    className="no-selectable"
                  >
                    {expansion}
                  </Text>
                  <SimpleGrid mb={3} minChildWidth="90px" spacing={2}>
                    {groupedCards[expansion].map(
                      ({ name, series, type, cost, description }) => (
                        <Tooltip
                          key={`card-${name}`}
                          hasArrow
                          label={
                            <>
                              <Text>
                                [{cost}] {type}
                              </Text>
                              <Text>{description}</Text>
                            </>
                          }
                          placement="top"
                          bg="gray.100"
                          color="black"
                          shadow={false}
                        >
                          <Center
                            height="50px"
                            border="1px"
                            color={
                              selections.includes(series) ? 'white' : 'black'
                            }
                            bgColor={
                              selections.includes(series)
                                ? submitted
                                  ? 'green.400'
                                  : 'blue.400'
                                : 'transparent'
                            }
                            borderColor={
                              selections.includes(series)
                                ? 'transparant'
                                : 'gray.200'
                            }
                            borderRadius={4}
                            textAlign="center"
                            cursor={
                              submitted ||
                              (selections.length === nSelections &&
                                !selections.includes(series))
                                ? 'auto'
                                : 'pointer'
                            }
                            onClick={(e) => {
                              selectCard(series)
                            }}
                            opacity={
                              selections.length === nSelections &&
                              !selections.includes(series)
                                ? 0.4
                                : 1
                            }
                          >
                            <Text fontSize="sm" className="no-selectable">
                              {Mikan.split(name).map((text, index) => (
                                <span
                                  key={`card-${name}-${index}`}
                                  className="no-break"
                                >
                                  {text}
                                </span>
                              ))}
                            </Text>
                          </Center>
                        </Tooltip>
                      )
                    )}
                  </SimpleGrid>
                </div>
              ))}
            </VStack>
            <HStack spacing={4}>
              <Link href={`/`}>トップページに戻る</Link>
              <Link onClick={clearUserName}>ユーザー名を変更する</Link>
              <Spacer />
            </HStack>
          </Layout>
        </>
      )}
      {!userName && (
        <Layout>
          <Text mb={2} fontSize="xl" fontWeight="600" textAlign="center">
            Room ID: {router.query.roomId} ({nUsers}人, {nSelections}枚)
          </Text>
          <HStack mb={12}>
            <Input
              value={userNameValue}
              onChange={handleUserNameValueChange}
              placeholder="あなたのユーザー名を決めてください"
            />
            <Button
              onClick={updateUserName}
              px={10}
              colorScheme="blue"
              disabled={userNameValue === ''}
            >
              決定
            </Button>
          </HStack>
        </Layout>
      )}
      {result && (
        <Modal
          isOpen={result}
          size="xl"
          closeOnEsc={false}
          closeOnOverlayClick={false}
          isCentered
        >
          <ModalOverlay />
          <ModalContent>
            <ModalHeader>集計結果</ModalHeader>
            <ModalBody>
              <UnorderedList>
                {result.logs.map((log, index) => (
                  <ListItem key={`result-log-${index}`}>{log}</ListItem>
                ))}
              </UnorderedList>
              <Box mt={4} mb={8} p={4} bgColor="gray.50">
                <UnorderedList>
                  {Object.keys(result.cards).map((expansion, index) => (
                    <ListItem key={`result-extension-${index}`}>
                      <span>{expansion}から</span>
                      {result.cards[expansion].map(({ name }) => (
                        <span key={`result-card-${name}`}>「{name}」</span>
                      ))}
                    </ListItem>
                  ))}
                </UnorderedList>
                <Text>を選択してゲームを開始してください。</Text>
              </Box>
            </ModalBody>
          </ModalContent>
        </Modal>
      )}
    </>
  )
}
